## import dependencies

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## load data

In [2]:
df = pd.read_csv('../data/raw.csv')
df.head()

ID_Pengiriman              Tanggal  Waktu_Keberangkatan  \
0     DEL-00001  2023-10-09 10:28:00  2023-10-09 10:28:00   
1     DEL-00002  2023-09-15 17:07:00  2023-09-15 17:07:00   
2     DEL-00003  2023-09-21 13:57:00  2023-09-21 13:57:00   
3     DEL-00004  2023-09-19 13:10:00  2023-09-19 13:10:00   
4     DEL-00005  2023-09-11 14:52:00  2023-09-11 14:52:00   

             Waktu_Kedatangan  Lat_Jemput  Lon_Jemput  Lat_Tujuan  Lon_Tujuan  \
0  2023-10-09 12:09:35.444386   -6.187770  106.884394   -6.324869  106.783430   
1  2023-09-15 17:53:45.354053   -6.140012  106.898952   -6.140302  106.711119   
2  2023-09-21 14:05:42.112794   -6.101567  106.948339   -6.204143  106.927419   
3  2023-09-19 13:53:23.355129   -6.112538  106.778596   -6.292112  106.655991   
4  2023-09-11 15:15:47.785700   -6.085218  106.912228   -6.172220  106.816521   

     Jarak_KM_Terukur  ID_Kurir Tipe_Kendaraan Pengalaman_Kurir  Jumlah_Paket  \
0   39.44772552617686        21          Motor         Menengah            37   
1  19.092557707822586         7          Motor         Menengah             5   
2                 1.0         1          Motor         Menengah             5   
3  23.393817976950878         1            MTR           Rendah            11   
4  17.144595029161525        18      Mobil_Van           Rendah            16   

  Kondisi_Cuaca Kondisi_Lalu_Lintas Tipe_Area_Tujuan  Durasi_Pengiriman_Menit  \
0         Badai              Sedang        Perumahan               101.590740   
1  Hujan_Ringan              Ringan        Perumahan                46.755901   
2         Cerah              Sedang        Perumahan                 8.701880   
3         Cerah              Sedang        Perumahan                43.389252   
4         Cerah              Ringan        Perumahan                23.796428   

   ETA_Awal_Menit  Status_Keterlambatan  
0       63.494212                     1  
1       42.690170                     0  
2        7.614145                     0  
3       37.965596                     0  
4       24.986250                     0

## Cleaning Data

In [7]:
# Cek struktur dan tipe data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 19 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID_Pengiriman            10000 non-null  object 
 1   Tanggal                  10000 non-null  object 
 2   Waktu_Keberangkatan      10000 non-null  object 
 3   Waktu_Kedatangan         10000 non-null  object 
 4   Lat_Jemput               10000 non-null  float64
 5   Lon_Jemput               10000 non-null  float64
 6   Lat_Tujuan               9800 non-null   float64
 7   Lon_Tujuan               10000 non-null  float64
 8   Jarak_KM_Terukur         10000 non-null  object 
 9   ID_Kurir                 10000 non-null  int64  
 10  Tipe_Kendaraan           10000 non-null  object 
 11  Pengalaman_Kurir         10000 non-null  object 
 12  Jumlah_Paket             10000 non-null  int64  
 13  Kondisi_Cuaca            10000 non-null  object 
 14  Kondisi_Lalu_Lintas    

1. tipe data yang salah
    - tanggal
    - waktu keberangkatan
    - waktu kedatangan
    - jarak km terukur
2. nilai kosong
    - lan tujuan
    - durasi pengiriman menit

In [4]:
# cek statistik deskriptif
df.describe()

Lat_Jemput    Lon_Jemput   Lat_Tujuan    Lon_Tujuan      ID_Kurir  \
count  10000.000000  10000.000000  9800.000000  10000.000000  10000.000000   
mean      -6.197676    106.798379    -6.200301    106.800355     24.933600   
std        0.086216      0.086299     0.086885      0.086432     14.105293   
min       -6.349986    106.650002    -6.349995    106.650003      1.000000   
25%       -6.271467    106.722851    -6.275431    106.725907     13.000000   
50%       -6.196705    106.798648    -6.200916    106.801431     25.000000   
75%       -6.124507    106.872333    -6.124130    106.874930     37.000000   
max       -6.050030    106.949937    -6.050008    106.949987     49.000000   

       Jumlah_Paket  Durasi_Pengiriman_Menit  ETA_Awal_Menit  \
count  10000.000000              9900.000000    10000.000000   
mean      21.858200                47.440540       40.337984   
std       10.057107                62.566147       24.940046   
min        5.000000                 5.000000      -21.089756   
25%       13.000000                22.292487       20.602444   
50%       22.000000                43.100462       40.027153   
75%       31.000000                64.924636       59.093225   
max       39.000000              1967.535417      119.756683   

       Status_Keterlambatan  
count          10000.000000  
mean               0.046500  
std                0.210576  
min                0.000000  
25%                0.000000  
50%                0.000000  
75%                0.000000  
max                1.000000

1 hari ada 1440 menit, tapi nilai maksimal durasi pengiriman menit itu 1967.5 berarti waktu pengiriman itu lebih dari 1 hari. terdapat potensi outlier di kolom durasi_pegiriman_menit, dilihat dari nilai Q3 dan max yang memilki perbedaan yang jauh

In [9]:
# cek nilai null
(df.isnull().sum()/len(df))*100

ID_Pengiriman              0.0
Tanggal                    0.0
Waktu_Keberangkatan        0.0
Waktu_Kedatangan           0.0
Lat_Jemput                 0.0
Lon_Jemput                 0.0
Lat_Tujuan                 2.0
Lon_Tujuan                 0.0
Jarak_KM_Terukur           0.0
ID_Kurir                   0.0
Tipe_Kendaraan             0.0
Pengalaman_Kurir           0.0
Jumlah_Paket               0.0
Kondisi_Cuaca              0.0
Kondisi_Lalu_Lintas        0.0
Tipe_Area_Tujuan           0.0
Durasi_Pengiriman_Menit    1.0
ETA_Awal_Menit             0.0
Status_Keterlambatan       0.0
dtype: float64

- null value pada kolom lat_tujuan sebanyak 2% dari seluruh data
- null value pada kolom durasi_pengiriman_menit sebanyak 1% dari seluruh data

In [10]:
# cek nilai duplikat
df.duplicated().sum()

np.int64(0)

tidak terdapat nilai duplikat

### Masalah pada data
1. tipe data tidak sesui : Tanggal, Waktu_Keberangkatan, Waktu_Kedatangan, Jarak_KM_Terukur
2. Nilai null : Lon_Tujuan (2% data), Durasi_Pengiriman_Menit(1% data)
3. Potensi nilai outlier : Durasi_Pengiriman_Menit

### Mengubah tipe data

In [11]:
# object ke datetime
df['Tanggal'] = pd.to_datetime(df['Tanggal'])
df['Waktu_Keberangkatan'] = pd.to_datetime(df['Waktu_Keberangkatan'])
df['Waktu_Kedatangan'] = pd.to_datetime(df['Waktu_Kedatangan'])


In [13]:
df['Jarak_KM_Terukur'] = df['Jarak_KM_Terukur'].astype(float)

ValueError: could not convert string to float: '5.85 KM'

terdapat data yang memiliki KM

In [14]:
df['Jarak_KM_Terukur'] = df['Jarak_KM_Terukur'].str.replace('KM', '', regex=False)

In [15]:
df['Jarak_KM_Terukur'] = df['Jarak_KM_Terukur'].str.strip() # menghapus spasi

In [16]:
df['Jarak_KM_Terukur'] = df['Jarak_KM_Terukur'].astype(float)

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 19 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   ID_Pengiriman            10000 non-null  object        
 1   Tanggal                  10000 non-null  datetime64[ns]
 2   Waktu_Keberangkatan      10000 non-null  datetime64[ns]
 3   Waktu_Kedatangan         10000 non-null  datetime64[ns]
 4   Lat_Jemput               10000 non-null  float64       
 5   Lon_Jemput               10000 non-null  float64       
 6   Lat_Tujuan               9800 non-null   float64       
 7   Lon_Tujuan               10000 non-null  float64       
 8   Jarak_KM_Terukur         10000 non-null  float64       
 9   ID_Kurir                 10000 non-null  int64         
 10  Tipe_Kendaraan           10000 non-null  object        
 11  Pengalaman_Kurir         10000 non-null  object        
 12  Jumlah_Paket             10000 no

### Mengatasi Nilai null
karena data yang kosong hanya 2% dan 1% maka data kosong tersebut akan di hapus.

In [18]:
df.dropna(inplace=True)

In [20]:
df.isnull().sum()

ID_Pengiriman              0
Tanggal                    0
Waktu_Keberangkatan        0
Waktu_Kedatangan           0
Lat_Jemput                 0
Lon_Jemput                 0
Lat_Tujuan                 0
Lon_Tujuan                 0
Jarak_KM_Terukur           0
ID_Kurir                   0
Tipe_Kendaraan             0
Pengalaman_Kurir           0
Jumlah_Paket               0
Kondisi_Cuaca              0
Kondisi_Lalu_Lintas        0
Tipe_Area_Tujuan           0
Durasi_Pengiriman_Menit    0
ETA_Awal_Menit             0
Status_Keterlambatan       0
dtype: int64

### Menangani potensi outlier

In [ ]:
# deteksi outlier dengan IQR
Q1 = df['Durasi_Pengiriman_Menit'].quantile(0.25)
Q3 = df['Durasi_Pengiriman_Menit'].quantile(0.75)
IQR = Q3-Q1

batas_bawah = Q1-1.5*IQR
batas_atas = Q3+1.5*IQR

outlier = df[(df['Durasi_Pengiriman_Menit'] < batas_bawah) | (df['Durasi_Pengiriman_Menit'] > batas_atas)]
outlier


ID_Pengiriman             Tanggal Waktu_Keberangkatan  \
27       DEL-00028 2023-09-06 12:41:00 2023-09-06 12:41:00   
65       DEL-00066 2023-09-02 16:44:00 2023-09-02 16:44:00   
716      DEL-00717 2023-09-15 09:10:00 2023-09-15 09:10:00   
851      DEL-00852 2023-09-29 15:38:00 2023-09-29 15:38:00   
889      DEL-00890 2023-10-14 12:25:00 2023-10-14 12:25:00   
...            ...                 ...                 ...   
9266     DEL-09267 2023-09-08 09:08:00 2023-09-08 09:08:00   
9394     DEL-09395 2023-09-16 16:34:00 2023-09-16 16:34:00   
9450     DEL-09451 2023-10-22 19:13:00 2023-10-22 19:13:00   
9699     DEL-09700 2023-10-12 07:42:00 2023-10-12 07:42:00   
9722     DEL-09723 2023-10-11 12:03:00 2023-10-11 12:03:00   

               Waktu_Kedatangan  Lat_Jemput  Lon_Jemput  Lat_Tujuan  \
27   2023-09-06 15:06:54.836308   -6.340599  106.685796   -6.327649   
65   2023-09-02 19:20:05.432852   -6.159998  106.896790   -6.319959   
716  2023-09-15 11:49:53.512158   -6.128199  106.729380   -6.311111   
851  2023-09-29 18:16:37.114305   -6.278378  106.671569   -6.325129   
889  2023-10-14 14:36:49.219776   -6.117799  106.747277   -6.286937   
...                         ...         ...         ...         ...   
9266 2023-09-08 11:51:31.769260   -6.064512  106.758498   -6.272039   
9394 2023-09-16 16:40:36.441577   -6.305369  106.904237   -6.341463   
9450 2023-10-22 21:24:37.684330   -6.337422  106.925249   -6.266557   
9699 2023-10-12 09:55:18.684312   -6.067475  106.945410   -6.112399   
9722 2023-10-11 14:49:31.967500   -6.169134  106.939318   -6.214757   

      Lon_Tujuan  Jarak_KM_Terukur  ID_Kurir Tipe_Kendaraan Pengalaman_Kurir  \
27    106.917911         46.978585        20          Motor         Menengah   
65    106.693164         51.568471         1      Mobil_Van           Rendah   
716   106.925318         47.335373         1          Motor         Menengah   
851   106.911746         41.397017        48          Motor         Menengah   
889   106.745302         40.426481        18            Van         Menengah   
...          ...               ...       ...            ...              ...   
9266  106.722392         50.245271        27      Mobil_Van         Menengah   
9394  106.701916          4.205039        19      Mobil_Van           Tinggi   
9450  106.719594         43.045496        22          Motor         Menengah   
9699  106.746914         45.036907        31          Motor         Menengah   
9722  106.657735         34.979702        22      Mobil_Van         Menengah   

      Jumlah_Paket Kondisi_Cuaca Kondisi_Lalu_Lintas Tipe_Area_Tujuan  \
27              11         Cerah               Padat        Perumahan   
65              20         Cerah               Padat           Bisnis   
716             24         Cerah               Padat        Perumahan   
851             10  Hujan_Ringan               Padat        Perumahan   
889             17         Cerah               Padat        Perumahan   
...            ...           ...                 ...              ...   
9266            15         Badai              Sedang         Industri   
9394            16         Badai              Ringan        Perumahan   
9450            36         Badai              Ringan        Perumahan   
9699            11  Hujan_Ringan               Padat         Industri   
9722            30         Badai               Padat        Perumahan   

      Durasi_Pengiriman_Menit  ETA_Awal_Menit  Status_Keterlambatan  
27                 145.913938       95.756022                     1  
65                 156.090548      102.434422                     1  
716                159.891869      104.929039                     1  
851                158.618572       90.516033                     1  
889                131.820330       86.507091                     1  
...                       ...             ...                   ...  
9266               163.529488      102.205930                     1  
9394     

In [27]:
len(outlier)/len(df)*100

0.9380476239562933

In [28]:
outlier['Durasi_Pengiriman_Menit'].min()

np.float64(128.9046819476653)

In [29]:
outlier['Durasi_Pengiriman_Menit'].max()

np.float64(1967.53541665578)

terdapat 0.93% data yang terindikasi outlier. tetapi jika dianalisis lebih lanjut, angka ini tergolong normal, karena kondisi lalu lintas yang tergolong padat atau kondisi cuaca yang badai. tetapi ada nilai 1935.105107 menit dengan kondisi lalu lintas yang ringan dan kondisi cuaca badai. Nilai tersebut dibandingan dengan baris yang lain dengan kondisi lalu lintas yang ringan dan cuaca yang padat tidak memiliki data Durasi_Pengiriman_Menit setinggi itu.

In [31]:
outlier.groupby('Kondisi_Lalu_Lintas')['Durasi_Pengiriman_Menit'].agg(['min', 'max', 'mean'])

min          max         mean
Kondisi_Lalu_Lintas                                      
Padat                128.904682   189.809055   144.583862
Ringan               131.628072  1944.945323  1552.110523
Sedang               129.002018  1967.535417   651.703492

In [33]:
outlier.groupby('Kondisi_Cuaca')['Durasi_Pengiriman_Menit'].agg(['min', 'max', 'mean'])

min          max        mean
Kondisi_Cuaca                                     
Badai          131.628072  1935.105107  359.897312
Cerah          129.457498  1967.535417  410.048440
Hujan_Ringan   128.904682  1944.945323  228.674079

terdapat pengiriman waktu yang lebih dari 1 hari atau 1440 menit

In [35]:
# melihat data yang lebih dari 1 hari
outlier[outlier['Durasi_Pengiriman_Menit']>1440]

ID_Pengiriman             Tanggal Waktu_Keberangkatan  \
1371     DEL-01372 2023-10-03 16:38:00 2023-10-03 16:38:00   
1670     DEL-01671 2023-10-25 12:53:00 2023-10-25 12:53:00   
2794     DEL-02795 2023-10-02 19:31:00 2023-10-02 19:31:00   
4148     DEL-04149 2023-10-18 10:14:00 2023-10-18 10:14:00   
4188     DEL-04189 2023-10-29 11:15:00 2023-10-29 11:15:00   
4764     DEL-04765 2023-09-13 12:04:00 2023-09-13 12:04:00   
4980     DEL-04981 2023-09-16 17:51:00 2023-09-16 17:51:00   
6364     DEL-06365 2023-09-29 07:19:00 2023-09-29 07:19:00   
6661     DEL-06662 2023-09-09 10:52:00 2023-09-09 10:52:00   
9394     DEL-09395 2023-09-16 16:34:00 2023-09-16 16:34:00   

               Waktu_Kedatangan  Lat_Jemput  Lon_Jemput  Lat_Tujuan  \
1371 2023-10-03 17:21:00.444606   -6.072578  106.692712   -6.309059   
1670 2023-10-25 12:58:00.000000   -6.194760  106.802556   -6.100845   
2794 2023-10-02 20:35:30.447536   -6.254026  106.947527   -6.056468   
4148 2023-10-18 12:02:27.224239   -6.285652  106.742031   -6.216228   
4188 2023-10-29 12:05:24.814613   -6.149241  106.774197   -6.270612   
4764 2023-09-13 12:24:57.380967   -6.308414  106.923416   -6.075185   
4980 2023-09-16 19:12:31.631858   -6.243589  106.870889   -6.325917   
6364 2023-09-29 07:42:51.667001   -6.112725  106.819172   -6.135980   
6661 2023-09-09 12:17:58.970366   -6.275194  106.821799   -6.204382   
9394 2023-09-16 16:40:36.441577   -6.305369  106.904237   -6.341463   

      Lon_Tujuan  Jarak_KM_Terukur  ID_Kurir Tipe_Kendaraan Pengalaman_Kurir  \
1371  106.681416         15.075906        13          Motor         Menengah   
1670  106.942769          1.000000         1      Mobil_Van           Tinggi   
2794  106.743568         38.577619        42      Mobil_Van         Menengah   
4148  106.689318         45.776071         9          Motor           Rendah   
4188  106.915840         28.345649        26      Mobil_Van           Tinggi   
4764  106.739116          9.638730        36          Motor         Menengah   
4980  106.921035         39.195793        45          Motor         Menengah   
6364  106.800921         17.231370        13          Motor         Menengah   
6661  106.778156         41.396154        13          Motor           Tinggi   
9394  106.701916          4.205039        19      Mobil_Van           Tinggi   

      Jumlah_Paket Kondisi_Cuaca Kondisi_Lalu_Lintas Tipe_Area_Tujuan  \
1371            20         Cerah              Sedang           Bisnis   
1670            27  Hujan_Ringan              Ringan        Perumahan   
2794            21         Cerah              Ringan           Bisnis   
4148            36         Badai              Ringan           Bisnis   
4188            10         Cerah              Ringan        Perumahan   
4764            17         Cerah              Sedang        Perumahan   
4980            35  Hujan_Ringan              Ringan           Bisnis   
6364             9         Cerah              Ringan         Industri   
6661            36         Cerah              Sedang           Bisnis   
9394            16         Badai              Ringan        Perumahan   

      Durasi_Pengiriman_Menit  ETA_Awal_Menit  Status_Keterlambatan  
1371              1967.535417       37.631484                     0  
1670              1671.503314        2.520306                     0  
2794              1699.568851       67.732832                     0  
4148              1512.566637       81.340303                     0  
4188              1709.857400       58.815840                     0  
4764              1902.744447       18.336806                     0  
4980              1944.945323       74.437876                     0  
6364              1811.709478       25.054173                     0  
6661              1668.211491       83.594427                     0  
9394              1935.105107        5.506133                     0

dari data diatas, kondisi cuaca tidak buruk atau ada 1 data yang buruk yaitu badan dan kurir mengguanakan mobil memiliki waktu pengiriman lebih dari 1 hari. data yang memiliki nilai lebih dari 1 hari akan dihapus

In [37]:
df = df[df['Durasi_Pengiriman_Menit']<1440]

In [40]:
df['Durasi_Pengiriman_Menit'].max()

np.float64(189.80905474976456)

In [44]:
df.select_dtypes(include='number').describe()

Lat_Jemput   Lon_Jemput   Lat_Tujuan   Lon_Tujuan  Jarak_KM_Terukur  \
count  9691.000000  9691.000000  9691.000000  9691.000000       9691.000000   
mean     -6.197638   106.798249    -6.200412   106.800504         21.289130   
std       0.086212     0.086386     0.086903     0.086352         12.615379   
min      -6.349986   106.650002    -6.349995   106.650003          1.000000   
25%      -6.271613   106.722333    -6.275658   106.726160         11.004318   
50%      -6.196669   106.798409    -6.201043   106.801456         21.132235   
75%      -6.124559   106.872642    -6.124198   106.874909         31.142623   
max      -6.050030   106.949937    -6.050008   106.949987         59.020737   

          ID_Kurir  Jumlah_Paket  Durasi_Pengiriman_Menit  ETA_Awal_Menit  \
count  9691.000000   9691.000000              9691.000000     9691.000000   
mean     24.965432     21.827262                45.698012       40.334227   
std      14.114640     10.048227                29.175116       24.953385   
min       1.000000      5.000000                 5.000000      -21.089756   
25%      13.000000     13.000000                22.264043       20.594671   
50%      25.000000     22.000000                43.032301       39.956525   
75%      37.000000     30.000000                64.850414       59.073355   
max      49.000000     39.000000               189.809055      119.756683   

       Status_Keterlambatan  
count           9691.000000  
mean               0.046538  
std                0.210658  
min                0.000000  
25%                0.000000  
50%                0.000000  
75%                0.000000  
max                1.000000